CosineDistance.lua

```diff
@@ -8,6 +8,9 @@ end
 function CosineDistance:updateOutput(input)
    local input1, input2 = input[1], input[2]
 
+   if not input1:isContiguous() then input1 = input1:clone() end
```

Comment:
Hi,
Could you use a temporary buffer to avoid memory allocations at every forward/backward pass ? Something like
```lua
if not input1:isContiguous() then
  self._input1 = self._input1 or input1.new()
  self._input1:resizeAs(input1):copy(input1)
  input1 = self._input1
end
```
You could reuse this `_input1`/`_input2` buffer in the backward pass as well.
Or maybe wrapping it in a function like https://github.com/torch/nn/blob/master/SpatialConvolution.lua#L65 ?
Thanks !